File: Jimin0304/week3/InsertionSort.py
# 공간복잡도 : 주어진 배열이 차지하고 있는 공간 내에서 값들의 위치만 바꾸기 때문에 O(1)
# 시간복잡도 :
# 루프문을 통해 정렬 범위를 2개로 시작해서 전체로 확장해야 하기 때문에 기본적으로 O(N)을 시간을 소모,
# 각 패스에서는 정렬 범위에 새롭게 추가된 값과 기존 값들의 대소 비교 및 자리 교대를 위해서 O(N)을 시간이 필요
# 즉 O(N^2)의 시간 복잡도를 가지는 정렬 알고리즘

# 최적화를 통해서 부분적으로 정렬된 배열에 대해서 성능을 대폭 개선할 수 있으며, 
# 특히 완전히 정렬되어 있는 배열이 들어올 경우, O(N)까지도 시간 복잡도를 향상시킬 수 있음

def insertion_sort(arr):
    num = len(arr)
    for i in range (1, num):
        key = arr[i]
        for j in range (i - 1, -1, -1):  #i-1부터 0까지 역순으로
            if (key < arr[j]):
                arr[j + 1], arr[j] = arr[j], key
        
arr = [7, 5, 1, 4, 3]
print("original:", arr)
insertion_sort(arr)
print("insertion sort:", arr)

File: heeeete/week3/insertion_sort.c
/*	시간복잡도 - O(n^2)
	공간복잡도 - O(1)
	선택정렬, 버블정렬에 비해 상대적으로 빠름	*/

#include <stdio.h>

void insertion_sort(int arr[], int N)
{
	int key, j;
	for(int i = 1 ; i < N ; i++)
	{
		key = arr[i];
		for(j = i - 1; j >= 0 && key < arr[j]; j--) //J는 0보다 같거나 커야하며 key 값이 arr[j]보다 크면 반복문 종료
		{
			arr[j + 1] = arr[j];
		}
		arr[j + 1] = key;
	}
}


/*	int main()
{
		int arr[] = {5,1,2,2,1,6,7,4,6,78};
		insertion_sort(arr, sizeof(arr) / sizeof(int));
		for(int i = 0 ; i < sizeof(arr) / sizeof(int) ; i++)
			printf("%d\n", arr[i]);	
}	*/